kratos/elements/levelset_convection_element_simplex_algebraic_stabilization.h

```diff
@@ -0,0 +1,434 @@
+
+            v[i] = GetGeometry()[i].FastGetSolutionStepValue(rConvVar);
+            vold[i] = GetGeometry()[i].FastGetSolutionStepValue(rConvVar,1);
+
+            X_mean_tmp += GetGeometry()[i].Coordinates();
+            X_node[i] = GetGeometry()[i].Coordinates();
+
+            grad_phi_mean_tmp += GetGeometry()[i].GetValue(rGradVar);
+
+            phi_mean_old += GetGeometry()[i].FastGetSolutionStepValue(rUnknownVar,1);
+            phi_mean += GetGeometry()[i].FastGetSolutionStepValue(rUnknownVar);
+        }
+
+        phi_mean /= static_cast<double>(TNumNodes);
+        phi_mean_old /= static_cast<double>(TNumNodes);
+
+        array_1d<double,TDim> X_mean, grad_phi_mean;
+        for(unsigned int k = 0; k < TDim; k++)
+        {
+            grad_phi_mean[k] = grad_phi_mean_tmp[k]/static_cast<double>(TNumNodes);
+            X_mean[k] = X_mean_tmp[k]/static_cast<double>(TNumNodes);
+        }
+
+        array_1d<double,TDim> grad_phi_diff = prod(trans(DN_DX), phi_old) - grad_phi_mean;
+
+        BoundedMatrix<double,TNumNodes, TNumNodes> K_matrix = ZeroMatrix(TNumNodes, TNumNodes); // convection
+        BoundedMatrix<double,TNumNodes, TNumNodes> S_matrix = ZeroMatrix(TNumNodes, TNumNodes); // LHS stabilization
+        Vector S_vector = ZeroVector(TNumNodes); // RHS stabilization
+        BoundedMatrix<double,TNumNodes, TNumNodes> Mc_matrix = ZeroMatrix(TNumNodes, TNumNodes); // consistent mass matrix
+        BoundedMatrix<double,TNumNodes, TNumNodes> Ml_matrix = ZeroMatrix(TNumNodes, TNumNodes); // lumped mass matrix
+
+        BoundedMatrix<double,TNumNodes, TNumNodes> Ncontainer;
+        GetShapeFunctionsOnGauss(Ncontainer);
+        for(unsigned int igauss=0; igauss<TDim+1; igauss++)
+        {
+            noalias(N) = row(Ncontainer,igauss);
+
+            //obtain the velocity/coordinate at the gauss point
+            array_1d<double, TDim > vel_gauss=ZeroVector(TDim);
+            array_1d<double, TDim > X_gauss=ZeroVector(TDim);
+            double phi_gauss = 0.0;
+            double phi_gauss_old = 0.0;
+            for (unsigned int i = 0; i < TNumNodes; i++)
+            {
+                for(unsigned int k=0; k<TDim; k++)
+                {
+                    vel_gauss[k] += N[i]*v[i][k];
+                    X_gauss[k] += N[i]*X_node[i][k];
+                }
+                phi_gauss += N[i]*phi[i];
+                phi_gauss_old += N[i]*phi_old[i];
+            }
+
+            array_1d<double, TNumNodes > v_dot_grad_N = prod(DN_DX, vel_gauss);
+
+            noalias(Mc_matrix) += outer_prod(N, N);
+            noalias(K_matrix) += outer_prod(N, v_dot_grad_N);
+
+            for (unsigned int i = 0; i < TNumNodes; i++){
+                S_vector[i] += ( (phi_gauss_old - phi_mean_old) - inner_prod(grad_phi_mean,(X_gauss - X_mean)) )*N[i];
+            }
+        }
+
+        for (unsigned int i = 0; i < TNumNodes; i++)
+        {
+            Ml_matrix(i,i) = 1.0;
+        }
```

Comment:
You can avoid this loop by initializing `Ml_matrix` with `IdentityMatrix` instead of `ZeroMatrix`.